adagrams/game.py

```diff
@@ -1,11 +1,139 @@
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+#helper function to get a random letter from LETTERS_ALPHABET dictionary
+def random_selected_letter():
+    letters_alphabet = list(LETTERS_ALPHABET.keys())
+    #randint() is part of random library and random_index will return a random index between 0 and the higher limit in the length of letters_alphabet list
+    #every time we use this function it will retrieve a random index number
+    random_index = random.randint(0, len(letters_alphabet)-1)
+    #we'll use that random_index to retrieve the element (letter) that is in that position in the list
+    random_letter = letters_alphabet[random_index]
+
+    return random_letter
+
 def draw_letters():
-    pass
+    #hand variable will storage all the 10 random letters
+    hand = []
+    random_letter = ""
+
+    #while loop to loop through both lists letters_alphabet and LETTERS_MAX_ALLOWED to make
+    # sure we don't use more than the maximun amount allowed per letter
+    # while loop stops when we have 10 letters appended  
+    while len(hand) < 10:
+        random_letter= random_selected_letter()
+        if hand.count(random_letter) < LETTERS_ALPHABET[random_letter]:
+            hand.append(random_letter)
 
+    return hand
+
+#helper function to convert a string or list into a dictionary with the amount of times that element appears
+#as it value
+def word_letter_dictionaries(data_to_convert):
+    dict_convert = {}
+
+    for elem in data_to_convert:
+        if elem in dict_convert:
+            dict_convert[elem] +=1
+        else:
+            dict_convert[elem]= 1
+    return dict_convert
+    
 def uses_available_letters(word, letter_bank):
-    pass
+    #valid variable to check whether the letter is in letter_bank or not
+    valid = False
+    word_upper = word.upper()
+    #dict_word and dict_letter_bank store a dictionary with each letter as key and the amount of that that letter
+    #appears as value
+    dict_word = word_letter_dictionaries(word_upper)
+    dict_letter_bank = word_letter_dictionaries(letter_bank)
+    #for loop to loop through dict_word  and check whether the letters are in dict_letter_bank
+    #or not and in case the amount of that letter exceeds the amount in dict_letter_bank it returns False
+    for elem in dict_word:
+        if elem not in letter_bank or dict_word[elem] > dict_letter_bank[elem]:
+            valid = False
+            break
+        else:
+            valid = True
+
+    return valid
 
 def score_word(word):
-    pass
+    #create dict_score to store letters that shares the same point value
+    dict_score = {1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+                  2: ["D", "G"],
+                  3: ["B", "C", "M", "P"],
+                  4: ["F", "H", "V", "W", "Y"],
+                  5: ["K"],
+                  8: ["J", "X"],
+                  10: ["Q", "Z"]}
```

Comment:
👀 By laying out the dictionary as score→letter list, this is fundamentally the same approach as what you already had. We have to iterate through the dictionary pairs, then iterate through each list to see whether this score applies to the current letter. This is a more data-driven approach, absolutely. But we should also keep in mind the strengths of a dictionary and the optimal way of arranging data in it.

When using a dictionary, be sure to use the data we have to look up (here each letter in the word) as the key, and the value we want to get as a result of the look up (the letter score) as the value. This leverages the strength of a dictionary (quick lookups by key).

If you find yourself iterating over a dictionary, take a moment to consider whether the data in the dictionary should be arranged differently.